pennylane/optimize/qnspsa.py

```diff
@@ -0,0 +1,407 @@
+        # the computational ket{0} state, which equals the state overlap between the perturbed and
+        # unperturbed circuits.
+        tensor_finite_diff = (
+            tensor_raw_results[0][0]
+            - tensor_raw_results[1][0]
+            - tensor_raw_results[2][0]
+            + tensor_raw_results[3][0]
+        )
+        metric_tensor = (
+            -(
+                np.tensordot(tensor_dirs[0], tensor_dirs[1], axes=0)
+                + np.tensordot(tensor_dirs[1], tensor_dirs[0], axes=0)
+            )
+            * tensor_finite_diff
+            / (8 * self.finite_diff_step * self.finite_diff_step)
+        )
+        return metric_tensor
+
+    def _get_next_params(self, args, gradient):
+        params = []
+        for arg in args:
+            if not getattr(arg, "requires_grad", False):
+                continue
+            # if an input parameter is a zero-dimension array, add one dimension to form an array.
+            # The returned result will not drop this added dimension
+            if arg.shape == ():
+                arg = arg.reshape(-1)
+            params.append(arg)
+
+        # params_vec and grad_vec group multiple inputs into the same vector to solve the
+        # linear equation
+        params_vec = np.concatenate(params).reshape(-1)
+        grad_vec = np.concatenate(gradient).reshape(-1)
+
+        new_params_vec = np.linalg.solve(
+            self.metric_tensor,
+            (-self.stepsize * grad_vec + np.matmul(self.metric_tensor, params_vec)),
+        )
+        # reshape single-vector new_params_vec into new_params, to match the input params
+        params_split_indices = []
+        tmp = 0
+        for param in params:
+            tmp += param.size
+            params_split_indices.append(tmp)
+        new_params = np.split(new_params_vec, params_split_indices)
+        new_params_reshaped = [new_params[i].reshape(params[i].shape) for i in range(len(params))]
+        return new_params_reshaped
+
+    def _get_spsa_grad_tapes(self, cost, args, kwargs):
+        dirs = []
+        args_plus = list(args)
+        args_minus = list(args)
+        for index, arg in enumerate(args):
+            if not getattr(arg, "requires_grad", False):
+                continue
+            direction = self.rng.choice([-1, 1], size=arg.shape)
+
+            dirs.append(direction)
+            args_plus[index] = arg + self.finite_diff_step * direction
+            args_minus[index] = arg - self.finite_diff_step * direction
+
+        cost.construct(args_plus, kwargs)
+        tape_plus = cost.tape.copy(copy_operations=True)
+        cost.construct(args_minus, kwargs)
+        tape_minus = cost.tape.copy(copy_operations=True)
+        return [tape_plus, tape_minus], dirs
+
+    def _update_tensor(self, tensor_raw):
+        def get_tensor_moving_avg(metric_tensor):
+            if self.metric_tensor is None:
+                self.metric_tensor = np.identity(metric_tensor.shape[0])
```

Comment:
Why not initialize `self.metric_tensor` in `self.__init__` instead of here?

Reply:
Following PennyLane's convention, the qnode that an optimizer will be applied on is not declared when initializing a new optimizer. 
```python
opt = qml.QNSPSAOptimizer(stepsize=stepsize)
new_params = opt.step(qnode, current_params)
```
We will then have no information about the dimension of this ``self.metric_tensor`` until the first optimization step. Let me know if this makes sense.

Reply:
Right, that makes sense 🙂